### DISCUSSION TITLE
Magic++: Expanded color map not written to file

### DISCUSSION THREAD

**Discussion by HorstBaerbel** [ImageMagick/ImageMagick6#336] [General]
Using Magic++ 6.9.7.4+dfsg-16ubuntu6.4 amd64 on Ubuntu 18.04.2.
When I load a 8-bit paletted PNG, resize the color map and write to the file again the color map still has the same size, not the expanded size as expected. This also happens if I I set the expanded, previously unused entries.
Example code:
```
Image img;
img.read("in.png"); //<-- 8bit paletted file, 255 colors
std::vector<Color> colorMap;
for (size_t i = 0; i < img->colorMapSize(); ++i) {
    colorMap.push_back(img->colorMap(i));
}
img.colorMapSize(img.colorMapSize() + 1); //<--after this this the color map is actually 1 entry bigger
colorMap.push_back(Color("#ff00ff"));
for (size_t i = 0; i < colorMap.size(); ++i) {
    img->colorMap(i, colorMap.at(i));
}
img.write("out.png"); //<--image written has still 255 colors
```

I also tried calling img.modifyImage() before and img.syncPixels () afterwards, but that didn't change anything. There's no need to expand the color map. A simple swap of color map entries refuses to work for me...

**Comment by HorstBaerbel** [comment]
Is this a bug, or am I "doing it wrong"? I am simply trying to add a color to the color map at index #0.

**Comment by dlemstra** [comment]
It might be possible that the PNG coder will remove the color unless it is used in the image. What happens when you set one of the pixels to that color?

**Comment by HorstBaerbel** [comment]
Haven't tried, but it is not usable workaround, as then you need to have a pixel you don't need in the image... :/
My use case is that I'm converting images for a paletted display system (GBA). There color No. 0 is always transparent, thus I need to add a color at palette index 0 that I don't use. My original code used was this:
```
Image img;
img.read("in.png"); //<-- 8bit paletted file, 255 colors
// read palette
std::vector<Color> colorMap;
for (size_t i = 0; i < img->colorMapSize(); ++i) {
    colorMap.push_back(img->colorMap(i));
}
img.colorMapSize(img.colorMapSize() + 1); //<--after this this the color map is actually 1 entry bigger
colorMap.push_back(Color("#ff00ff"));
for (size_t i = 0; i < colorMap.size(); ++i) {
    img->colorMap(i, colorMap.at(i));
}
img.syncPixels();
// increase color indices in pixels
img.modifyImage();
const auto nrOfIndices = img.columns() * img.rows();
auto pixels = img.getPixels(0, 0, img.columns(), img.rows()); // we need to call this first for getIndexes to work
auto indices = img.getIndexes();
for (size_t i = 0; i < nrOfIndices; ++i) {
    indices[i] = indices[i] + 1;
}
// sync and write modified image
img.syncPixels();
img.write("out.png"); //<--image written has still 255 colors
```
But this doesn't work either.
